rfcs/NNN-form-associated-mixin.md

```diff
@@ -0,0 +1,377 @@
+  to define validation.
+- This RFC does not aim to cover all form-related element use cases, such as
+  defining a form and managing form state. It it solely concerned with helpers
+  for defining individual form-associated elements, which could be used with
+  form management utilities.
+
+## Motivation
+
+Building a form-associated custom element requires a lot of common boilerplate:
+
+- Setting `static formAssociated = true`
+- Attaching `ElementInternals`
+- Setting a default ARIA role
+- Calling `internals.setFormValue()` when needed
+- Calling `internals.setValidity()` and `internals.checkValidity()` when needed
+- Handling `formDisabledCallback()`
+- Handling `formResetCallback()` and `formStateRestoreCallback()`
+- Requesting a reactive update when state held in internals changes.
+
+This boilplate makes it much harder to implement a form-associated element.
+Element authors should be able to focus on the actually necessary and unique
+parts of their element:
+
+- The element's value and state
+- Validation logic
+
+## Detailed Design
+
+A new `@lit-labs/forms` package will contain several utilities:
+- A `FormAssociated()` mixin that:
+  - Makes an element form associated
+  - Implements the standard form associated custom element callbacks
+  - Provides a validation hook
+  - Adds no public APIs to the element
+- A set of decorators for designating fields as the form value and state
+  - `@formValue()` for the public form value field
+  - `@formState()` for the fields that compose part of the form state
+  - `@formDefaultValue()` for the public form default value field
+  - `@formDefaultState()` for the public form default state field
+  - `@formStateGettter()` for private serializion of the form state
+  - `@formStateSetter()` for private deserializion of the form state
+- A `FormControll()` mixin that implements common public APIs for form controls,
+  like `.form`, `.disabled`, `.validity`, etc.
+
+### Why a mixin?
+
+`FormAssociated` needs to be a mixin because we need to implement the form
+associated callbacks on the element class. Controllers do not work for this use
+case.
+
+### Why decorators?
+
+A key choice in this design is to not require or add any _public_ API of the implementing element. Rather than automatically adding `.value` or `.disabled`
+fields, we leave that up to the element authors.
+
+In order for the mixin to function, it needs to perform some actions and read
+some element state when form-relate state changes. Decorators let us wrap
+arbitrary element accessors to invoke the mixin code paths to update form state,
+and let us communicate hooks to the mixin declaratively.
+
+```ts
+class MyFormElement extends FormAssociated(LitElement) {
+
+  // Setting this field calls internals.setFormValue()
+  @formValue()
+  @property({reflect: true})
+  accessor value: string = '';
+
+  // When the form resets, the value is set to this field
+  @formDefaultValue()
+  accessor defaultValue: string = '';
```

Comment:
Is this representative of how normal form controls operate? In my (limited) testing, `value` is usually _not_ reflected, and the current value for the `value` attribute is used as default, so I expected:

```ts
  // Setting this field calls internals.setFormValue()
  @formValue()
  @property({reflect: false, attribute: false})
  accessor value: string = '';

  // When the form resets, the value is set to this field
  @formDefaultValue()
  @property({attribute: 'value'})
  accessor defaultValue: string = '';

```

Reply:
The point here is that you can use the decorators to construct the public API you want, including as you point out, one that matches the oddities of `<input>`. Yes, you do not have to reflect any of these decorated fields.